### PR TITLE
Feat auto round

### PR DESCRIPTION
**Pull request by pablomlago**
## Reason for this PR

Implement AutoRound within Brevitas (see https://github.com/intel/auto-round, https://arxiv.org/pdf/2309.05516)

## Changes Made in this PR

Incorporate AutoRound, refactored learned round methods into a single common interface.

## Testing Summary

New tests for the learned round utilities, replicate results from AutoRound repo.

## Risk Highlight

- [ ] This PR includes code from another work (please detail).
- [ ] This PR contains API-breaking changes.
- [ ] This PR depends on work in another PR (please provide links/details).
- [ ] This PR introduces new dependencies (please detail).
- [ ] There are coverage gaps not covered by tests.
- [ ] Documentation updates required in subsequent PR.

## Checklist

- [x] Code comments added to any hard-to-understand areas, if applicable.
- [x] Changes generate no new warnings.
- [x] Updated any relevant tests, if applicable.
- [x] No conflicts with destination `dev` branch.
- [x] I reviewed my own code changes.
- [ ] Initial CI/CD passing.
- [ ] 1+ reviews given, and any review issues addressed and approved.
- [ ] Post-review full CI/CD passing.

### CODE REVIEW

**Comment by nickfraser** [comment]
@pablomlago, can you switch this target `dev` not `master`?

File: src/brevitas/core/function_wrapper/auto_round.py
Comment (Giuseppe5) [inline src/brevitas/core/function_wrapper/auto_round.py:1]: We should have already a file with learned round imlementations. Let's put all in there

File: src/brevitas/optim/sign_sgd.py
Comment (Giuseppe5) [inline src/brevitas/optim/sign_sgd.py:1]: Is this the only file here?

File: src/brevitas/optim/sign_sgd.py [hunk 1]
```diff
@@ -0,0 +1,489 @@
+"""
```
Comment (Giuseppe5) [inline src/brevitas/optim/sign_sgd.py:1]: Why this license?

File: src/brevitas/optim/sign_sgd.py [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+            if group["momentum"] != 0:
+                # update momentum_buffers in state
+                for p, momentum_buffer in zip(params, momentum_buffer_list):
+                    state = self.state[p]
+                    state["momentum_buffer"] = momentum_buffer
+
+        return loss
+
+
+SignSGD.__doc__ = (
```
Comment (Giuseppe5) [inline src/brevitas/optim/sign_sgd.py:184]: Weird, if we want to have a docstring, let's use the same style as the rest of the repo (check some class in brevitas/core)

File: src/brevitas/optim/sign_sgd.py [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+        The Nesterov version is analogously modified.
+
+        Moreover, the initial value of the momentum buffer is set to the
+        gradient value at the first step. This is in contrast to some other
+        frameworks that initialize it to all zeros.
+
+    """)
+
+
+def sign_sgd(
```
Comment (Giuseppe5) [inline src/brevitas/optim/sign_sgd.py:276]: If this function (and above) are taken from somewhere, we need to specify from where

File: src/brevitas/optim/sign_sgd.py [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+                torch._foreach_add_(device_params, grads_x_lr)
+            else:
+                torch._foreach_add_(device_params, torch._foreach_sign(device_grads), alpha=-lr)
+        else:
+            # foreach APIs don't support sparse
+            for i in range(len(device_params)):
+                device_params[i].add_(torch.sign(device_grads[i]), alpha=-lr)
+
+
+def _fused_sign_sgd(
```
Comment (Giuseppe5) [inline src/brevitas/optim/sign_sgd.py:474]: What is this supposed to do?
If we are not planning to implement it, let's remove it.

File: src/brevitas_examples/llm/benchmark/llm_benchmark.py
Comment (Giuseppe5) [inline src/brevitas_examples/llm/benchmark/llm_benchmark.py:1]: This could be a separate PR

File: src/brevitas_examples/llm/llm_quant/learned_round_utils.py
Comment (Giuseppe5) [inline src/brevitas_examples/llm/llm_quant/learned_round_utils.py:1]: If anything is shared between this and imagenet, let's put it in one file and avoid duplication

**Review by Giuseppe5 — COMMENTED** [review]
Preliminary review, I'll play with the code a bit in the meantime while you address these comments

File: src/brevitas_examples/llm/benchmark/llm_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    def __hash__(self):
+        return hash(tuple(sorted(self.items())))
+
+
+def unique(sequence):
+    seen = set()
+    return [x for x in sequence if not (x in seen or seen.add(x))]
+
+
+# Torchvision models with top1 accuracy
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/benchmark/llm_benchmark.py:57]: Comment is wrong

File: src/brevitas/optim/sign_sgd.py (hunk 4 above)
Comment (pablomlago) [reply src/brevitas/optim/sign_sgd.py:396]: This would be used in a high-performance implementation, in which some operations are fused. Probably, this is not needed now, but I left the signature in case it is required at some point in the future.

File: src/brevitas_examples/common/learned_round/learned_round_method.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+        total_loss = rec_loss + round_loss
+        return total_loss, (total_loss, rec_loss, round_loss, b)
+
+    def format_loss_components(self, loss: float, rec_loss: float, round_loss: float, b) -> str:
+        return "loss = {:.4f}, rec_loss = {:.4f}, round_loss = {:.4f}, b = {:.4f}".format(
+            loss, rec_loss, round_loss, b)
+
+
+class AdaRound(LearnedRound):
```
Comment (Giuseppe5) [inline src/brevitas_examples/common/learned_round/learned_round_method.py:150]: Let's rename this, not sure to what

File: src/brevitas_examples/common/learned_round/learned_round_method.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+
+    def __call__(self, pred: torch.Tensor, tgt: torch.Tensor) -> Tuple[torch.Tensor, Tuple]:
+        loss = F.mse_loss(pred, tgt)
+        return loss, (loss,)
+
+    def format_loss_components(self, loss: float) -> str:
+        return "loss = {:.4f}".format(loss)
+
+
+class AutoRound(LearnedRound):
```
Comment (Giuseppe5) [inline src/brevitas_examples/common/learned_round/learned_round_method.py:209]: Same as above, rename to something that makes more clear the difference between this FloatToIntImpl and the one above.

File: src/brevitas_examples/llm/main.py [hunk 8]
```diff
@@ -285,6 +288,17 @@ def main(args):
 
     model = offload_model(model)
 
+    if args.learned_round:
+        print("Applying learned round...")
+
+        learned_round_llm_utils = LearnedRoundLLMUtils()
+        learned_round = AutoRound()
+        learned_round_optimiser = LearnedRoundOptimizer(
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/main.py:296]: Not a fan of this entrypoint, I'd rather have strings that have a meaning, that are interpreted within the class to the correct classes

File: src/brevitas_examples/llm/main.py [hunk 9]
```diff
@@ -285,6 +288,17 @@ def main(args):
 
     model = offload_model(model)
 
+    if args.learned_round:
+        print("Applying learned round...")
+
+        learned_round_llm_utils = LearnedRoundLLMUtils()
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/main.py:294]: Ideally, we shouldn't be needing this class, and try to merge everything with the vision one.
For anything that can't be merged, it is user defined.

File: tests/brevitas/optim/test_sign_sgd.py [hunk 10]
```diff
@@ -0,0 +1,310 @@
+"""
```
Comment (Giuseppe5) [inline tests/brevitas/optim/test_sign_sgd.py:1]: Still confused about this license (not sure where my previous comment went)

File: tests/brevitas_examples/test_imagenet.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
+from hypothesis import given
```
Comment (Giuseppe5) [inline tests/brevitas_examples/test_imagenet.py:1]: Missing license

File: tests/brevitas_examples/test_learned_round_utils.py [hunk 12]
```diff
@@ -0,0 +1,325 @@
+from hypothesis import given
```
Comment (Giuseppe5) [inline tests/brevitas_examples/test_learned_round_utils.py:1]: Missing license

File: src/brevitas_examples/common/learned_round/learned_round_method.py [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+
+    def __call__(self, t):
+        if t < self.start_decay:
+            return self.start_b
+        else:
+            rel_t = (t - self.start_decay) / (self.t_max - self.start_decay)
+            return self.end_b + (self.start_b - self.end_b) * max(0.0, (1 - rel_t))
+
+
+class AdaRoundLoss(LearnedRoundLoss):
```
Comment (Giuseppe5) [inline src/brevitas_examples/common/learned_round/learned_round_method.py:102]: Change name to something more meaningful

File: src/brevitas/core/function_wrapper/learned_round.py [hunk 14]
```diff
@@ -52,6 +53,7 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         return p
 
 
+# TODO: Change name to AdaRoundSte for consistency
```
Comment (Giuseppe5) [inline src/brevitas/core/function_wrapper/learned_round.py:56]: Nope don't change

File: src/brevitas_examples/llm/llm_quant/learned_round_utils.py [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+        if disable_quant:
+            disable_quant_class.enable_act_quantization(model, False)
+            disable_quant_class.enable_param_quantization(model, False)
+            restore_return_quant_tensor(model, return_quant_tensor_state)
+
+    def init_model_learned_round(self, model: nn.Module) -> None:
+        self.llm_cache_state = model.config.use_cache
+        model.config.use_cache = False
+
+    def finish_model_learned_round(self, model: nn.Module) -> None:
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/llm_quant/learned_round_utils.py:108]: This would be a part of some function implemented in llm utils file, not of the PTQ algorithm

Comment (Giuseppe5) [reply src/brevitas_examples/llm/llm_quant/learned_round_utils.py:108]: This method is not needed in that case

File: src/brevitas_examples/llm/llm_quant/learned_round_utils.py [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+                    model(**inps)
+                except StopFwdException:
+                    pass
+        handle.remove()
+        if disable_quant:
+            disable_quant_class.enable_act_quantization(model, False)
+            disable_quant_class.enable_param_quantization(model, False)
+            restore_return_quant_tensor(model, return_quant_tensor_state)
+
+    def init_model_learned_round(self, model: nn.Module) -> None:
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/llm_quant/learned_round_utils.py:104]: Same as below

File: src/brevitas_examples/llm/llm_quant/learned_round_utils.py [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+        self.llm_cache_state = model.config.use_cache
+        model.config.use_cache = False
+
+    def finish_model_learned_round(self, model: nn.Module) -> None:
+        model.config.use_cache = self.llm_cache_state
+        self.llm_cache_state = None
+
+    def init_cache(self) -> Any:
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/llm_quant/learned_round_utils.py:112]: Same as above

File: src/brevitas_examples/llm/llm_quant/learned_round_utils.py [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+        outs = torch.cat([cache_outs[i] for i in indices], dim=input_dim)
+        # Make sure that the inputs and outputs are in the same device as block,
+        # before running its forward pass.
+        args = send_to_device(args, device)
+        kwargs = send_to_device(kwargs, device)
+        outs = send_to_device(outs, device)
+
+        return (args, kwargs), outs
+
+    def run_forward(
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/llm_quant/learned_round_utils.py:198]: This shouldn't be here. Rather, the user should be able to specify whatever interface for the input to the model.
We should provide the interface to accept a function with a certain signature, and the user decides what happens inside that function.

Possible signature
```python
def model_forward(model, model_args, model_kwargs):
```

File: src/brevitas_examples/llm/llm_quant/learned_round_utils.py [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+        quant_outs = block(*args, **kwargs)
+        if isinstance(quant_outs, tuple):
+            quant_outs = quant_outs[0]
+        return quant_outs
+
+    def loss_scaler(
+        self,
+        loss: torch.Tensor,
+    ) -> torch.Tensor:
+        return loss * 1000
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/llm_quant/learned_round_utils.py:213]: Hardcoded stuff, bad.
Why is this here?

File: src/brevitas_examples/llm/llm_quant/learned_round_utils.py [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+class LearnedRoundLLMUtils(LearnedRoundModelUtils):
+
+    def __init__(self) -> None:
+        super(LearnedRoundLLMUtils, self).__init__()
+        self.llm_cache_state = None
+
+    def default_block_check_fn(self, module: nn.Module, module_name: str) -> bool:
+        return isinstance(module, LlamaDecoderLayer) or isinstance(module, OPTDecoderLayer)
+
+    class _DataSaverHookLLM:
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/llm_quant/learned_round_utils.py:29]: Is this different from the CNN version? How?

File: src/brevitas/optim/sign_sgd.py (hunk 2 above)
Comment (pablomlago) [reply src/brevitas/optim/sign_sgd.py:136]: Note that this docstring is not only literals, but contains some variables, so it cannot be set in the same style as others of the repo. However, it should be checked that this plays nicely with the Sphinx documentation. If not, it might be worth removing the variables and just have a literal docstring.

File: tests/brevitas/optim/test_sign_sgd.py (hunk 10 above)
Comment (pablomlago) [reply tests/brevitas/optim/test_sign_sgd.py:1]: Similar to: https://github.com/Xilinx/brevitas/blob/master/src/brevitas/nn/quant_mha.py.

File: src/brevitas/optim/sign_sgd.py (hunk 1 above)
Comment (pablomlago) [reply src/brevitas/optim/sign_sgd.py:1]: Similar to https://github.com/Xilinx/brevitas/blob/master/src/brevitas/nn/quant_mha.py.

File: src/brevitas_examples/llm/main.py (hunk 8 above)
Comment (pablomlago) [reply src/brevitas_examples/llm/main.py:296]: Agreed. I created a builder method. I don't think it's optimal still, and I'd be happy to iterate over it.

File: src/brevitas_examples/llm/main.py (hunk 9 above)
Comment (pablomlago) [reply src/brevitas_examples/llm/main.py:294]: My idea with this design is to decouple the logic of the PTQ algorithm, which is contained in LearnedRoundOptimizer, from anything that it is model/dataloader specific, e.g. capturing inputs/outputs for a block. This needs to be managed by specific utilities, which adhere to the interface LearnedRoundModelUtils (thus following the strategy pattern). By doing so, a potential change in the vision models would not break the LLM entrypoint, and viceversa.

File: src/brevitas/core/function_wrapper/learned_round.py [hunk 21]
```diff
@@ -92,3 +94,36 @@ def _load_from_state_dict(
         value_key = prefix + 'value'
         if config.IGNORE_MISSING_KEYS and value_key in missing_keys:
             missing_keys.remove(value_key)
+
+
+class AutoRoundSte(brevitas.jit.ScriptModule):
```
[the remark this answers is not in the document]
Comment (pablomlago) [reply src/brevitas/core/function_wrapper/learned_round.py:99]: Currently we have different float_to_int_impl (round, ceil, floor, ...), each of them with their corresponding STE (RoundSte, FloorSte, ...). Therefore, it seems sensible to me for those implementations of float_to_int_impl which involve learned parameters to be given the same treatment, and not aiming to aggregate them within a single general class, while keeping separate classes for those rounding methods which are not learnable. Moreover, the learned round methods that we have right now only have a single parameter tensor, but this does not need to be the case for future methods, so by trying to aggregated the learned round methods under a common umbrella now, we might make it more difficult in the future to integrate other methods.

File: src/brevitas_examples/llm/llm_quant/learned_round_utils.py (hunk 18 above)
Comment (pablomlago) [reply src/brevitas_examples/llm/llm_quant/learned_round_utils.py:198]: This signature is defined in LearnedRoundModelUtils.

File: src/brevitas/optim/sign_sgd.py (hunk 3 above)
Comment (Giuseppe5) [reply src/brevitas/optim/sign_sgd.py:198]: Yes

File: src/brevitas/optim/sign_sgd.py (hunk 4 above)
Comment (Giuseppe5) [reply src/brevitas/optim/sign_sgd.py:396]: I doubt that it's a concern for us. Remove, maybe let's open an issue

File: src/brevitas_examples/common/learned_round/learned_round_builder.py [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+        return LearnedRoundVisionUtils
+    elif utils_type == "llm":
+        return LearnedRoundLLMUtils
+    else:
+        raise Exception(f"Learned round utilities for {utils_type} are not recognized.")
+
+
+def solve_learned_round_method_cls(method_type) -> LearnedRound:
+    if method_type == "ada_round":
+        return AdaRound
```
Comment (Giuseppe5) [inline src/brevitas_examples/common/learned_round/learned_round_builder.py:27]: No AdaRound/AutoRound.

LearnedRound with options Sigmoid, HardSigmoid, Linear

File: src/brevitas/core/function_wrapper/learned_round.py (hunk 14 above)
Comment (Giuseppe5) [reply src/brevitas/core/function_wrapper/learned_round.py:56]: Remove comment

File: src/brevitas_examples/llm/llm_quant/learned_round_utils.py [hunk 23]
```diff
@@ -185,30 +72,84 @@ def sample_cache(
...
+        out = out[0]
+    return out
+
+
+def llm_block_check_fn(module: nn.Module, module_name: str) -> bool:
+    return isinstance(module, LlamaDecoderLayer) or isinstance(module, OPTDecoderLayer)
+
+
+def apply_learned_round(model: nn.Module, calibration_loader: DataLoader) -> None:
+    iters = 200
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/llm_quant/learned_round_utils.py:116]: All these parameters must be exposed as well

File: src/brevitas/core/scaling/standalone.py [hunk 24]
```diff
@@ -241,17 +241,17 @@ def __init__(
         self.value = Parameter(torch.full(scaling_shape, 1.0, dtype=dtype, device=device))
 
     @brevitas.jit.script_method
-    def forward(self, ignored: Tensor, threshold: Optional[Tensor] = None) -> Tensor:
+    def forward(self, x: Tensor, threshold: Optional[Tensor] = None) -> Tensor:
```
Comment (Giuseppe5) [inline src/brevitas/core/scaling/standalone.py:244]: This looks off

File: src/brevitas/core/stats/stats_wrapper.py [hunk 25]
```diff
@@ -96,8 +97,12 @@ def __init__(
         super(_ParameterListStats, self).__init__()
 
         self.stats_input_concat_dim = stats_input_concat_dim
-        self.first_tracked_param = _ViewParameterWrapper(
-            tracked_parameter_list[0], stats_input_view_shape_impl)
+        if len(tracked_parameter_list) >= 1:
```
Comment (Giuseppe5) [inline src/brevitas/core/stats/stats_wrapper.py:100]: As before, these changes are not from this PR, what's happening?

File: src/brevitas_examples/llm/llm_quant/learned_round_utils.py [hunk 26]
```diff
@@ -0,0 +1,226 @@
...
+from torch import nn
+from torch.optim.lr_scheduler import LinearLR
+from torch.optim.lr_scheduler import LRScheduler
+from torch.optim.optimizer import Optimizer
+from torch.utils.data.dataloader import DataLoader
+from transformers.models.llama.modeling_llama import LlamaDecoderLayer
+from transformers.models.opt.modeling_opt import OPTDecoderLayer
+
+from brevitas.optim.sign_sgd import SignSGD
+from brevitas_examples.common.learned_round.learned_round_method import AutoRound
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/llm_quant/learned_round_utils.py:18]: No more autoround

File: tests/brevitas_examples/test_learned_round_utils.py [hunk 27]
```diff
@@ -0,0 +1,327 @@
...
+        block = quant_model.in_proj_mlp
+        learned_round.insert_learned_round_quantizers(block)
+
+        for module in block.modules():
+            if hasattr(module, "weight_quant"):
+                assert module.weight_quant.rounding_mode == "LEARNED_ROUND"
+                assert isinstance(
+                    module.weight_quant.tensor_quant.int_quant.float_to_int_impl, LearnedRoundSte)
+
+    @pytest.mark.parametrize("learned_round", [AutoRound(), AdaRound()])
```
Comment (Giuseppe5) [inline tests/brevitas_examples/test_learned_round_utils.py:316]: No more adaround/autoround

File: src/brevitas/core/function_wrapper/learned_round.py [hunk 28]
```diff
@@ -24,14 +25,18 @@ class LearnedRoundHardSigmoid(brevitas.jit.ScriptModule):
...
     def __init__(self, learned_round_zeta: float = 1.1, learned_round_gamma: float = -0.1) -> None:
         super(LearnedRoundHardSigmoid, self).__init__()
+        self.float_to_int_ste = floor_ste
+        self.is_p_value = True
         self.learned_round_zeta = learned_round_zeta
         self.learned_round_gamma = learned_round_gamma
 
     @brevitas.jit.script_method
-    def forward(self, x: torch.Tensor) -> torch.Tensor:
+    def forward(self, x: torch.Tensor, training: bool) -> torch.Tensor:
```
Comment (Giuseppe5) [inline src/brevitas/core/function_wrapper/learned_round.py:34]: Training is not needed. Check self.training

File: src/brevitas/core/function_wrapper/learned_round.py [hunk 29]
```diff
@@ -44,14 +49,34 @@ class LearnedRoundSigmoid(brevitas.jit.ScriptModule):
     def __init__(self, learned_round_temperature: float = 1.) -> None:
         super(LearnedRoundSigmoid, self).__init__()
         assert learned_round_temperature != 0, 'Temperature should be different than 0'
+        self.float_to_int_ste = floor_ste
+        self.is_p_value = True
         self.learned_round_temperature = learned_round_temperature
 
     @brevitas.jit.script_method
-    def forward(self, x: torch.Tensor) -> torch.Tensor:
+    def forward(self, x: torch.Tensor, training: bool) -> torch.Tensor:
```
Comment (Giuseppe5) [inline src/brevitas/core/function_wrapper/learned_round.py:57]: As before, self.training

File: src/brevitas_examples/common/learned_round/learned_round_method.py [hunk 30]
```diff
@@ -0,0 +1,186 @@
...
+from torch import nn
+import torch.nn.functional as F
+
+from brevitas.core.function_wrapper.learned_round import LearnedRoundSte
+from brevitas.inject.enum import FloatToIntImplType
+from brevitas.inject.enum import LearnedRoundImplType
+from brevitas.nn.quant_layer import QuantWeightBiasInputOutputLayer as QuantWBIOL
+
+
+class StopFwdException(Exception):
```
Comment (Giuseppe5) [inline src/brevitas_examples/common/learned_round/learned_round_method.py:18]: I'm sure we define this somewhere else, maybe GPTQ. Let's merge into one and put it in torch_utils.py

File: src/brevitas_examples/llm/llm_quant/learned_round_utils.py [hunk 31]
```diff
@@ -0,0 +1,226 @@
...
+
+from brevitas.inject.enum import LearnedRoundImplType
+from brevitas.optim.sign_sgd import SignSGD
+from brevitas_examples.common.learned_round.learned_round_method import LearnedRound
+from brevitas_examples.common.learned_round.learned_round_method import LearnedRoundLoss
+from brevitas_examples.common.learned_round.learned_round_method import MSELoss
+from brevitas_examples.common.learned_round.learned_round_optimizer import LearnedRoundOptimizer
+
+LEARNED_ROUND_MAP = {
+    "linear_round": LearnedRoundImplType.IDENTITY,}
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/llm_quant/learned_round_utils.py:25]: Why this? I still want to be able to select any. Similar for the other dicts

File: src/brevitas_examples/llm/llm_quant/learned_round_utils.py [hunk 32]
```diff
@@ -0,0 +1,226 @@
...
+    learned_round_optimizer.apply_learned_round(
+        model=model,
+        model_forward=llm_forward,
+        block_forward=llm_block_forward,
+        data_loader=calibration_loader,
+        cache=cache,
+        block_check_fn=llm_block_check_fn,
+        model_prepare_fn=llm_learned_round_prepare_fn,
+        model_finish_fn=llm_learned_round_finish_fn,
+        keep_gpu=False,
```
Comment (Giuseppe5) [inline src/brevitas_examples/llm/llm_quant/learned_round_utils.py:225]: Is this still not fixed?

File: src/brevitas_examples/common/learned_round/learned_round_optimizer.py [hunk 33]
```diff
@@ -0,0 +1,820 @@
...
+        self.lr_scheduler_kwargs = {} if lr_scheduler_kwargs is None else lr_scheduler_kwargs
+        self.lr_scheduler_kwargs["total_iters"] = self.iters
+
+        learned_round_loss_kwargs = {} if learned_round_loss_kwargs is None else learned_round_loss_kwargs
+        self.learned_round_loss_init = partial(
+            learned_round_loss_class, **learned_round_loss_kwargs)
+
+        # TODO: Remove once validated and expose the flag
+        # self.use_accelerate = use_accelerate
+        self.use_accelerate = False
```
Comment (Giuseppe5) [inline src/brevitas_examples/common/learned_round/learned_round_optimizer.py:410]: What does this mean

File: src/brevitas_examples/common/learned_round/learned_round_optimizer.py [hunk 34]
```diff
@@ -0,0 +1,820 @@
...
+        cache.initialize_cache()
+
+        # Iterate over blocks and optimise the rounding parameters within each of them
+        for block_idx, block in enumerate(blocks):
+            # Distribute the model across devices to run a forward pass to capture
+            # inputs/outputs to the given block
+            model = offload_model(model)
+            # Cache needs to be cleared before populating it with the inputs and outputs
+            # to the block under optimization.
+            cache.clear_cache()
```
Comment (Giuseppe5) [inline src/brevitas_examples/common/learned_round/learned_round_optimizer.py:741]: Fixed memory leak?

File: src/brevitas_examples/common/learned_round/learned_round_optimizer.py [hunk 35]
```diff
@@ -0,0 +1,820 @@
...
+                    params.requires_grad = False
+            for params in scale_params:
+                params.requires_grad = False
+
+            # Move the block back to CPU
+            block.cpu()
+
+            # TODO: This call might not be needed, check_clear and reset_cache methods
+            # Reset cache after optimisation
+            cache.reset_cache()
```
Comment (Giuseppe5) [inline src/brevitas_examples/common/learned_round/learned_round_optimizer.py:816]: Duplicated?

File: src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py [hunk 36]
```diff
@@ -26,195 +26,200 @@
...
-    pass
+def is_resnet_block(module: nn.Module, module_name: str) -> bool:
+    return (re.search(r"layer\d+", module_name) is not None)
+
+
+def is_layer(module: nn.Module, module_name: str) -> bool:
+    return isinstance(module, QuantWBIOL)
+
+
+LEARNED_ROUND_MAP = {
```
Comment (Giuseppe5) [inline src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py:61]: Duplicated of LLM one. In both cases, all options should be available to mix and match. Let's put it in one shared location.

File: src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py [hunk 37]
```diff
@@ -26,195 +26,200 @@
...
+    "sign_sgd": SignSGD,}
+BLOCK_CHECK_MAP = {
+    "layerwise": is_layer,
+    "blockwise": is_resnet_block,}
+LR_SCHEDULER_MAP = {
+    "linear": LinearLR,}
 
 
-class DataSaverHook:
+class CacheCNN(dict):
```
Comment (Giuseppe5) [inline src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py:78]: CacheVision

File: src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py [hunk 38]
```diff
@@ -26,195 +26,200 @@
...
-        disable_quant_class.enable_act_quantization(model, False)
-        disable_quant_class.enable_param_quantization(model, False)
-        restore_return_quant_tensor(model, return_quant_tensor_state)
-    return cached
+    img, _ = inputs
+    img = send_to_device(img, device)
+    model(img)
+
+
+def cnn_block_forward(block: nn.Module, inputs: Any) -> torch.Tensor:
```
Comment (Giuseppe5) [inline src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py:146]: There are minimal differences between this and the previous func. Can we have only one?

File: src/brevitas_examples/imagenet_classification/ptq/ptq_evaluate.py [hunk 39]
```diff
@@ -158,16 +158,44 @@ def validate_args(args):
...
+    choices=[None, 'linear'],
+    help='Learning rate scheduler for learned round (default: None)')
 parser.add_argument(
     '--learned-round-lr',
     default=1e-3,
     type=float,
     help='Learning rate for learned round (default: 1e-3)')
+parser.add_argument(
+    '--learned-round-batch-size',
+    default=1,
```
Comment (Giuseppe5) [inline src/brevitas_examples/imagenet_classification/ptq/ptq_evaluate.py:196]: Is this consistent with past behaviour?

File: src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py [hunk 40]
```diff
@@ -26,195 +26,200 @@
...
+
+    if optimizer not in OPTIMIZER_MAP:
+        raise ValueError(f"Optimizer {optimizer} is not available.")
+    optimizer_class = OPTIMIZER_MAP[optimizer]
+
+    if lr_scheduler is not None and lr_scheduler not in LR_SCHEDULER_MAP:
+        raise ValueError(f"Learning rate scheduler {lr_scheduler} is not available.")
+    lr_scheduler_class = None if lr_scheduler is None else LR_SCHEDULER_MAP[lr_scheduler]
+
+    optimizer_classes = {"adam": torch.optim.Adam, "sign_sgd": SignSGD}
```
Comment (Giuseppe5) [inline src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py:187]: Any particular reason why Adam? In other tests I've played a bit with SGD + momentum

Comment (Giuseppe5) [reply src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py:187]: Maybe we can have a string that is used to fetch whatever optimizer the user wants, as lonog as it's an official pytorch one. i.e. `getattr(torch.optim, optimizer)`
The drawback of this is how to handle the various kwargs for each optimizer

File: src/brevitas_examples/llm/benchmark/llm_benchmark.py
Comment (Giuseppe5) [inline src/brevitas_examples/llm/benchmark/llm_benchmark.py:1]: This must be another PR.

File: src/brevitas_examples/llm/benchmark/post_processing.py
Comment (Giuseppe5) [inline src/brevitas_examples/llm/benchmark/post_processing.py:1]: As above.

File: src/brevitas_examples/common/learned_round/learned_round_optimizer.py (hunk 33 above)
Comment (pablomlago) [reply src/brevitas_examples/common/learned_round/learned_round_optimizer.py:406]: I started the distributed training optimization. Maybe it's worth removing it and leave it for another PR.

File: src/brevitas_examples/imagenet_classification/ptq/ptq_evaluate.py (hunk 39 above)
Comment (pablomlago) [reply src/brevitas_examples/imagenet_classification/ptq/ptq_evaluate.py:196]: Yes, previously the learned round batch size was fixed to 1

File: src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py [hunk 41]
```diff
@@ -26,195 +26,170 @@
...
+    def collate_fn(self, batch: Any) -> Any:
+        raise NotImplementedError("This method is still not available for CNNs.")
+
+    def __len__(self):
+        return (
+            len(self["inputs"])
+            if isinstance(self["inputs"], list) else self["inputs"].shape[self.batch_dim])
+
+
+def cnn_forward(model: nn.Module, inputs: Any) -> None:
```
Comment (Giuseppe5) [inline src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py:123]: vision forward

File: src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py (hunk 38 above)
Comment (Giuseppe5) [reply src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py:130]: What about this?

File: src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py [hunk 42]
```diff
@@ -26,195 +26,170 @@
...
+from brevitas_examples.common.learned_round.learned_round_parser import parse_lr_scheduler_class
+from brevitas_examples.common.learned_round.learned_round_parser import parse_optimizer_class
 
 config.IGNORE_MISSING_KEYS = True
 
 
-class StopFwdException(Exception):
-    """Used to throw and catch an exception to stop traversing the graph."""
-    pass
+def is_resnet_block(module: nn.Module, module_name: str) -> bool:
```
[the remark this answers is not in the document]
Comment (pablomlago) [reply src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py:55]: But the user would need to specify a regular expression in this case, no?

[the remark this answers is not in the document]
Comment (Giuseppe5) [reply src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py:55]: Yes.

File: src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py [hunk 43]
```diff
@@ -26,195 +26,176 @@
...
 # SOFTWARE.
 
-import numpy as np
+import functools
+import re
+from typing import Any, Callable, Dict, Optional, Tuple, Union
+import warnings
+
+from accelerate.utils.operations import send_to_device
+from datasets import Dataset
```
Comment (Giuseppe5) [inline src/brevitas_examples/imagenet_classification/ptq/learned_round_utils.py:35]: Unused

File: src/brevitas_examples/common/learned_round/learned_round_parser.py [hunk 44]
```diff
@@ -0,0 +1,95 @@
+# Copyright (C) 2024, Advanced Micro Devices, Inc. All rights reserved.
+# SPDX-License-Identifier: BSD-3-Clause
+
+import re
+from typing import Any, Callable, Dict, Optional, Tuple, Type, Union
+import warnings
+
+from accelerate.utils.operations import send_to_device
+from datasets import Dataset
```
Comment (Giuseppe5) [inline src/brevitas_examples/common/learned_round/learned_round_parser.py:9]: Unused

File: src/brevitas_examples/common/learned_round/learned_round_parser.py [hunk 45]
```diff
@@ -0,0 +1,86 @@
...
+LEARNED_ROUND_MAP = {
+    "linear_round": LearnedRoundImplType.IDENTITY,
+    "hard_sigmoid_round": LearnedRoundImplType.HARD_SIGMOID,
+    "sigmoid_round": LearnedRoundImplType.SIGMOID,}
+LEARNED_ROUND_LOSS_MAP = {
+    "mse": MSELoss,
+    "regularised_mse": RegularisedMSELoss,}
+OPTIMIZER_MAP = {
+    "sign_sgd": SignSGD,}
+LR_SCHEDULER_MAP = {}
```
Comment (Giuseppe5) [inline src/brevitas_examples/common/learned_round/learned_round_parser.py:27]: This should not be empty, since we define one (LinearLR)

Comment (pablomlago) [reply src/brevitas_examples/common/learned_round/learned_round_parser.py:27]: That dictionary is only for custom LR Schedulers, default PyTorch ones are retrieved from torch.optim.lr_scheduler by matching the string passed by the user.
